lib/spack/spack/install_test.py

```diff
@@ -52,6 +94,640 @@ def get_test_stage_dir():
     )
 
 
+def cache_extra_test_sources(pkg: Pb, srcs: ListOrStringType):
+    """Copy relative source paths to the corresponding install test subdir
+
+    This routine is intended as an optional install test setup helper for
+    grabbing source files/directories during the installation process and
+    copying them to the installation test subdirectory for subsequent use
+    during install testing.
+
+    Args:
+        pkg: package being tested
+        srcs: relative path for file(s) and or subdirectory(ies) located in
+            the staged source path that are to be copied to the corresponding
+            location(s) under the install testing directory.
+
+    Raises:
+        OSError: if any of the source paths do not exist under the build stage
+        ValueError: if any of the source paths are absolute
+    """
+    paths = [srcs] if isinstance(srcs, str) else srcs
+    if any(os.path.isabs(path) for path in paths):
+        raise ValueError("Only relative paths can be used to copy test sources")
+
+    for path in paths:
+        src_path = os.path.join(pkg.stage.source_path, path)
+        dest_path = os.path.join(install_test_root(pkg), path)
+        if os.path.isdir(src_path):
+            fs.install_tree(src_path, dest_path)
+        elif os.path.exists(src_path):
+            fs.mkdirp(os.path.dirname(dest_path))
+            fs.copy(src_path, dest_path)
+        else:
+            raise OSError(f"Source path ('{path}') for the copy does not exist")
```

Comment:
This loop is a good candidate for collecting errors in advance and reporting all of them at the end (if any).